### PR TITLE
Initial openEBS support. Minor fixes in README

### PR DESCRIPTION
**Pull request by v-bpastiu**

### CODE REVIEW

File: DOKS-wordpress/README.md [hunk 1]
```diff
@@ -150,6 +152,255 @@ Finally, you need to setup the trusted sources between your MySQL database and y
...
+
+```console
+helm repo add openebs-nfs https://openebs.github.io/dynamic-nfs-provisioner
+
+helm repo update
+```
+
+Next, create a YAML file `openEBS-nfs-provisioner.yaml` to override the default helm values:
+
+```yaml
```
Comment (v-ctiutiu) [inline DOKS-wordpress/README.md:186]: Please avoid embedding the full values file in the README. It's big and hard to follow. 

Then, at a later time and if it's relevant for the guide, you can provide important sections from the values file and highlight the important settings.

File: DOKS-wordpress/README.md [hunk 2]
```diff
@@ -150,6 +152,255 @@ Finally, you need to setup the trusted sources between your MySQL database and y
...
+
+analytics:
+  enabled: "true"
+```
+
+**Note:**
+
+Most of the overrides are self-explanatory and can be customized. Please visit [openebs nfs provisioner helm values](https://github.com/openebs/dynamic-nfs-provisioner/blob/develop/deploy/helm/charts/values.yaml) for more details.
+
+Finally, install the chart using Helm:
```
Comment (v-ctiutiu) [inline DOKS-wordpress/README.md:335]: Before installing the chart, you need to tell the user to clone the repo (if not already), and change directory context (or provide the absolute path to the Helm values file).

File: DOKS-wordpress/README.md [hunk 3]
```diff
@@ -150,6 +152,255 @@ Finally, you need to setup the trusted sources between your MySQL database and y
...
+helm install openebs-nfs openebs-nfs/nfs-provisioner --version 0.9.0 \
+  --namespace openebs \
+  --create-namespace \
+  -f "openEBS-nfs-provisioner.yaml"
+```
+
+**Note:**
+A specific version for the Helm chart is used. In this case 0.9.0 was picked, which maps to the 0.9.0 version of the application. It’s good practice in general, to lock on a specific version. This helps to have predictable results, and allows versioning control via Git.
+
+You can verify Nginx deployment status via:
```
Comment (v-ctiutiu) [inline DOKS-wordpress/README.md:347]: What has Nginx to do with OpenEBS ?

File: DOKS-wordpress/README.md [hunk 4]
```diff
@@ -150,6 +152,255 @@ Finally, you need to setup the trusted sources between your MySQL database and y
...
+```
+
+The output looks similar to (notice that the STATUS column value is deployed):
+
+```text
+NAME            NAMESPACE       REVISION        UPDATED                                 STATUS          CHART                   APP VERSION
+openebs-nfs     openebs         1               2022-05-09 10:58:14.388721 +0300 EEST   deployed        nfs-provisioner-0.9.0   0.9.0  
+```
+
+NFS provisioner requires a block storage device to create the disk capacity required for the NFS server. Here, we will use the default Storage Class (do-block-storage) of the DigitalOcean Kubernetes cluster as the backend storage for the NFS provisioner. In that case, whichever application uses the newly created following Storage Class, can consume shared storage (NFS) on a DigitalOcean volume using OpenEBS NFS provisioner. The following steps create an OpenEBS NFS provisioner supported Storage Class.
```
Comment (v-ctiutiu) [inline DOKS-wordpress/README.md:360]: Avoid using `we`, `our`, etc.

File: DOKS-wordpress/README.md [hunk 5]
```diff
@@ -266,6 +522,87 @@ NAME                                   DESIRED   CURRENT   READY   AGE
 replicaset.apps/wordpress-6f55c9ffbd   1         1         1       23h
 ```
 
+Verify the PVSs created under the wordpress namespace and associated OpenEBS volume under the openebs namespace:
```
Comment (v-ctiutiu) [inline DOKS-wordpress/README.md:525]: PVSs - PVCs ?

File: DOKS-wordpress/README.md [hunk 6]
```diff
@@ -266,6 +522,87 @@ NAME                                   DESIRED   CURRENT   READY   AGE
...
+
+```console
+kubectl get pvc -A
+```
+
+The ouput looks similar to (notice the `RWX` access mode for the wordpress PVC and it's storage class created earlier when configuring the openEBS NFS provisioner):
```
Comment (v-ctiutiu) [inline DOKS-wordpress/README.md:531]: Maybe reformulate to:

>The output looks similar to (notice the `RWX` access mode for the wordpress PVC, and the new storage class defined earlier via the openEBS NFS provisioner):

File: DOKS-wordpress/README.md [hunk 7]
```diff
@@ -266,6 +522,87 @@ NAME                                   DESIRED   CURRENT   READY   AGE
...
+
+The ouput looks similar to:
+
+```text
+NAME                                       CAPACITY   ACCESS MODES   RECLAIM POLICY   STATUS   CLAIM                                                  STORAGECLASS       REASON   AGE
+pvc-b505c0af-e6ab-4623-8ad1-1bad784261d5   5Gi        RWX            Delete           Bound    wordpress/wordpress                                    rwx-storage                 23m
+pvc-d3f0c597-69ba-4710-bd7d-ed29ce41ce04   5Gi        RWO            Delete           Bound    openebs/nfs-pvc-b505c0af-e6ab-4623-8ad1-1bad784261d5   do-block-storage            23m
+```
+
+We can also create additional pod replicas to demonstrate the power of the NFS provisioner by opening the `values.yaml` file and uncommenting the: `replicaCount: 3` line (this can be set to as many replicas as you wish).
```
Comment (v-ctiutiu) [inline DOKS-wordpress/README.md:553]: Avoid using `we`, `our`, etc. And, maybe reformulate `demonstrate the power` to `demonstrate the capabilities`.

File: DOKS-wordpress/README.md [hunk 8]
```diff
@@ -266,6 +522,87 @@ NAME                                   DESIRED   CURRENT   READY   AGE
...
+    --values values.yml
+```
+
+Verify that the changes are applied:
+
+```console
+kubectl get all -n wordpress
+```
+
+The output looks similar to:
```
Comment (v-ctiutiu) [inline DOKS-wordpress/README.md:572]: Please also include an explanation about what the user should see (such as increased replica count, and number of pods).

File: DOKS-wordpress/README.md [hunk 9]
```diff
@@ -266,6 +522,87 @@ NAME                                   DESIRED   CURRENT   READY   AGE
...
+replicaset.apps/wordpress-798789f994   0         0         0       19m
+```
+
+We can also check where the pods are deployed:
+
+```console
+kubectl get all -n wordpress -o wide
+```
+
+The output looks similar to (notice that the pods are deployed in different nodes). The pods access the same Shared NFS Volume created earlier due to it's `RWX` or `ReadWriteMany` access mode.
```
Comment (v-ctiutiu) [inline DOKS-wordpress/README.md:597]: Maybe reformulate to:

> The output looks similar to (notice that the pods are deployed on different nodes):

File: DOKS-wordpress/README.md [hunk 10]
```diff
@@ -266,6 +522,87 @@ NAME                                   DESIRED   CURRENT   READY   AGE
...
+
+```text
+NAME                             READY   STATUS    RESTARTS   AGE     IP             NODE            NOMINATED NODE   READINESS GATES
+pod/wordpress-5f5f4cf94c-d7mqb   1/1     Running   0          4m7s    10.244.0.206   basicnp-cwxop   <none>           <none>
+pod/wordpress-5f5f4cf94c-qkxdq   1/1     Running   0          4m47s   10.244.1.84    basicnp-cwxol   <none>           <none>
+pod/wordpress-5f5f4cf94c-zf46h   1/1     Running   0          2m36s   10.244.0.194   basicnp-cwxop   <none>           <none>
+```
+
```
Comment (v-ctiutiu) [inline DOKS-wordpress/README.md:605]: And now, you can add the following sentence to continue the above:

>You can also observe pods accessing the same Shared NFS Volume created earlier, due to the `RWX` or `ReadWriteMany` flag.

File: DOKS-wordpress/README.md [hunk 11]
```diff
@@ -372,7 +710,7 @@ Apply via kubectl:
 kubectl apply -f letsencrypt-issuer.yaml
 ```
 
-To secure the traffic for WordPress, open the helm file `(value.yml)` created earlier and add the following helm values at the end:
+To secure the traffic for WordPress, open the helm file `(values.yml)` created earlier and add the following helm values at the end:
```
Comment (v-ctiutiu) [inline DOKS-wordpress/README.md:713]: Maybe reformulate to:

>To secure WordPress traffic, open the helm values file `(values.yml)` created earlier, and add the following settings at the end:

File: DOKS-wordpress/assets/manifests/values.yml [hunk 12]
```diff
@@ -33,18 +38,4 @@ externalDatabase:
...
-  tls: false
-  hostname: <YOUR_WORDPRESS_DOMAIN_HERE>
-  annotations:
-    kubernetes.io/ingress.class: "nginx"
-    cert-manager.io/cluster-issuer: "letsencrypt-prod"
-  extraTls:
-  - hosts:
-      - <YOUR_WORDPRESS_DOMAIN_HERE>
-    secretName: wordpress.local-tls
+  enabled: false
```
Comment (v-ctiutiu) [inline DOKS-wordpress/assets/manifests/values.yml:41]: Make sure that files have new line endings. There's a setting in VS Code which automatically adds a new line to files after saving. 
Open VS Code -> Preferences -> Settings, and then in the top bar search for `insert final newline`:

Screenshot 2022-05-11 at 09 35 45

File: DOKS-wordpress/README.md [hunk 13]
```diff
@@ -150,6 +152,255 @@ Finally, you need to setup the trusted sources between your MySQL database and y
...
+  annotations: 
+    openebs.io/cas-type: nsfrwx
+    cas.openebs.io/config: |
+      - name: NSFServerType
+        value: "kernel"
+      - name: BackendStorageClass
+        value: "do-block-storage"
+provisioner: openebs.io/nfsrwx
+reclaimPolicy: Delete
+```
```
Comment (v-ctiutiu) [inline DOKS-wordpress/README.md:379]: Usually, we also add some explanations for the relevant fields from the configuration manifest.

**Review by v-ctiutiu — CHANGES_REQUESTED** [review]

File: DOKS-wordpress/README.md [hunk 14]
```diff
@@ -150,6 +152,255 @@ Finally, you need to setup the trusted sources between your MySQL database and y
 
 Please visit [How to Secure MySQL Managed Database Clusters](https://docs.digitalocean.com/products/databases/mysql/how-to/secure/) for more details.
 
+## Introduction to the OpenEBS Dynamic NFS Provisioner
```
[the remark this answers is not in the document]
Comment (v-bpastiu) [reply DOKS-wordpress/README.md:155]: Will do.

File: DOKS-wordpress/README.md [hunk 15]
```diff
@@ -150,6 +151,143 @@ Finally, you need to setup the trusted sources between your MySQL database and y
...
+The output looks similar to:
+
+```text
+NAME                         PROVISIONER                 RECLAIMPOLICY   VOLUMEBINDINGMODE   ALLOWVOLUMEEXPANSION   AGE
+do-block-storage (default)   dobs.csi.digitalocean.com   Delete          Immediate           true                   24h
+```
+
+DigitalOcean Block Storage Volumes are mounted as read-write by a single node (RWO). Additional nodes cannot mount the same volume. The data content of a PersistentVolume can not be accessed by multiple Pods simultaneously.
+
+Horizontal pod autoscaling (HPA) policies are used to scale the WordPress Pods in a dynamically StatefulSet.
```
Comment (v-ctiutiu) [inline DOKS-wordpress/README.md:173]: Maybe combine these 2 sentences:

>Horizontal pod autoscaling (HPA) policies are used to scale the WordPress Pods in a dynamically StatefulSet.
>
>This is why WordPress requires a [volume](https://kubernetes.io/docs/concepts/storage/volumes/) that can be mounted as read-write by many nodes (RWX), in your case, WordPress replica Pods.

into:

>Horizontal pod autoscaling (HPA) is used to scale the WordPress Pods in a dynamically StatefulSet, hence WordPress requires a [volume](https://kubernetes.io/docs/concepts/storage/volumes/) mounted as read-write by many nodes (RWX).

Comment (v-ctiutiu) [inline DOKS-wordpress/README.md:173]: Maybe combine these 2 sentences:

>Horizontal pod autoscaling (HPA) policies are used to scale the WordPress Pods in a dynamically StatefulSet.
>
>This is why WordPress requires a [volume](https://kubernetes.io/docs/concepts/storage/volumes/) that can be mounted as read-write by many nodes (RWX), in your case, WordPress replica Pods.

into:

>Horizontal pod autoscaling (HPA) is used to scale the WordPress Pods in a dynamically StatefulSet, hence WordPress requires a [volume](https://kubernetes.io/docs/concepts/storage/volumes/) mounted as read-write by many nodes (RWX).

File: DOKS-wordpress/README.md [hunk 16]
```diff
@@ -150,6 +151,143 @@ Finally, you need to setup the trusted sources between your MySQL database and y
...
+
+This is why WordPress requires a [volume](https://kubernetes.io/docs/concepts/storage/volumes/) that can be mounted as read-write by many nodes (RWX), in your case, WordPress replica Pods.
+
+NFS (Network File System) is a commonly used solution to provide RWX volumes on block storage. This server offers a PersistentVolumeClaim (PVC) in RWX mode so that multiple web applications can access the data in a shared fashion.
+
+OpenEBS Dynamic NFS Provisioner allows users to create an NFS PV that sets up a new Kernel NFS instance for each PV on top of the user's choice of backend storage.
```
Comment (v-ctiutiu) [inline DOKS-wordpress/README.md:179]: > to create a NFS PV

Comment (v-ctiutiu) [inline DOKS-wordpress/README.md:179]: > to create a NFS PV

File: DOKS-wordpress/README.md [hunk 17]
```diff
@@ -150,6 +151,143 @@ Finally, you need to setup the trusted sources between your MySQL database and y
...
+
+**Note:**
+
+Please visit [OpenEBS](https://openebs.io/) for more details.
+
+Next, you will install the OpenEBS Dynamic NFS Provisioner on your Kubernetes cluster using the [OpenEBS Helm Chart](https://github.com/openebs/dynamic-nfs-provisioner). For your purposes you will be installing and configuring only the dynamic nfs provisioner since that's what you need.
```
Comment (v-ctiutiu) [inline DOKS-wordpress/README.md:185]: This sentence begins a little bit odd for me:

> For your purposes you will be installing and configuring only the dynamic nfs provisioner since that's what you need.

I think it can be written as:

>You will be installing and configuring only the dynamic nfs provisioner, since Wordpress requires it.

Comment (v-ctiutiu) [inline DOKS-wordpress/README.md:185]: >You will be installing and configuring only the dynamic nfs provisioner, since Wordpress requires it.

File: DOKS-wordpress/README.md [hunk 18]
```diff
@@ -150,6 +151,143 @@ Finally, you need to setup the trusted sources between your MySQL database and y
...
+Next, open and inspect the `assets/manifests/openEBS-nfs-provisioner.yaml` file provided in the repository:
+
+```yaml
+nfsStorageClass:
+  backendStorageClass: "do-block-storage"
+```
+
+**Note:**
+
+The override is self-explanatory and can be customized. Please visit [openebs nfs provisioner helm values](https://github.com/openebs/dynamic-nfs-provisioner/blob/develop/deploy/helm/charts/values.yaml) for the full values.yaml file and more details.
```
Comment (v-ctiutiu) [inline DOKS-wordpress/README.md:213]: Please avoid as much as possible constructs like: `self explanatory`, `it's obvious that`, `it's very simple to do something`, etc. Main point here is that readers may not be so knowledgeable as you are. So, in general even if some things `are obvious`, we should give a brief explanation for the reader.

Comment (v-ctiutiu) [inline DOKS-wordpress/README.md:213]: Please avoid as much as possible constructs like: `self explanatory`, `it's obvious that`, `it's very simple to do something`, etc. Main point here is that the user may not be so knowledgeable as you are. So, in general even if some things `are obvious`, we should explain it the user briefly.

File: DOKS-wordpress/README.md [hunk 19]
```diff
@@ -150,6 +151,143 @@ Finally, you need to setup the trusted sources between your MySQL database and y
...
+```
+
+The output looks similar to (notice that the STATUS column value is deployed):
+
+```text
+NAME            NAMESPACE       REVISION        UPDATED                                 STATUS          CHART                   APP VERSION
+openebs-nfs     openebs         1               2022-05-09 10:58:14.388721 +0300 EEST   deployed        nfs-provisioner-0.9.0   0.9.0  
+```
+
+NFS provisioner requires a block storage device to create the disk capacity required for the NFS server. Here, you will use the default Storage Class (do-block-storage) of the DigitalOcean Kubernetes cluster as the backend storage for the NFS provisioner. In that case, whichever application uses the newly created following Storage Class, can consume shared storage (NFS) on a DigitalOcean volume using OpenEBS NFS provisioner. The following steps create an OpenEBS NFS provisioner supported Storage Class.
```
Comment (v-ctiutiu) [inline DOKS-wordpress/README.md:240]: Maybe reformulate this sentence:

>Here, you will use the default Storage Class (`do-block-storage`) of the DigitalOcean Kubernetes cluster as the backend storage for the NFS provisioner.

To:

>Next, you will configure the default Kubernetes Storage Class (do-block-storage) provided by DigitalOcean as the backend storage for the NFS provisioner.

Comment (v-ctiutiu) [inline DOKS-wordpress/README.md:240]: Maybe reformulate this sentence:

>Here, you will use the default Storage Class (do-block-storage) of the DigitalOcean Kubernetes cluster as the backend storage for the NFS provisioner.

To:

>Next, you will configure the default Kubernetes Storage Class (do-block-storage) provided by DigitalOcean as the backend storage for the NFS provisioner.

File: DOKS-wordpress/README.md [hunk 20]
```diff
@@ -150,6 +151,143 @@ Finally, you need to setup the trusted sources between your MySQL database and y
...
+        value: "kernel"
+      - name: BackendStorageClass
+        value: "do-block-storage"
+provisioner: openebs.io/nfsrwx
+reclaimPolicy: Delete
+```
+
+Explanations for the above configuration:
+
+- `provisioner` - provisioner that determines what volume plugin is used for provisioning PVs (e.g. `openebs.io/nfsrwx`)
```
Comment (v-ctiutiu) [inline DOKS-wordpress/README.md:263]: Maybe:

>- `provisioner` - defines what storage class is used for provisioning PVs (e.g. `openebs.io/nfsrwx`)

Comment (v-ctiutiu) [inline DOKS-wordpress/README.md:263]: Maybe:

>- `provisioner` - defines what plugin is used for provisioning PVs (e.g. `openebs.io/nfsrwx`)

File: DOKS-wordpress/README.md [hunk 21]
```diff
@@ -150,6 +151,143 @@ Finally, you need to setup the trusted sources between your MySQL database and y
...
+The ouput looks similar to:
+
+```text
+NAME                         PROVISIONER                 RECLAIMPOLICY   VOLUMEBINDINGMODE   ALLOWVOLUMEEXPANSION   AGE
+do-block-storage (default)   dobs.csi.digitalocean.com   Delete          Immediate           true                   107m
+openebs-kernel-nfs           openebs.io/nfsrwx           Delete          Immediate           false                  84m
+rwx-storage                  openebs.io/nfsrwx           Delete          Immediate           false                  84m
+```
+
+Now, you have created the StorageClass named rwx-storage to dinamically privision shared volume on top of DigitalOcean Block Storage Volume.
```
Comment (v-ctiutiu) [inline DOKS-wordpress/README.md:289]: Maybe:

>Now, you have a new StorageClass named `rwx-storage` to dynamically provision shared volumes on top of DigitalOcean Block Storage.

Comment (v-ctiutiu) [inline DOKS-wordpress/README.md:289]: Maybe:

>Now, you have created the StorageClass named `rwx-storage` to dynamically provision shared volumes on top of the DigitalOcean Block Storage.

File: DOKS-wordpress/README.md [hunk 22]
```diff
@@ -266,6 +407,94 @@ NAME                                   DESIRED   CURRENT   READY   AGE
...
+replicaset.apps/wordpress-798789f994   0         0         0       19m
+```
+
+We can also check where the pods are deployed:
+
+```console
+kubectl get all -n wordpress -o wide
+```
+
+The output looks similar to (notice that the pods are deployed in different nodes):
```
Comment (v-ctiutiu) [inline DOKS-wordpress/README.md:489]: `in different nodes` -> `on different nodes`

Comment (v-ctiutiu) [inline DOKS-wordpress/README.md:489]: `on different nodes`

File: DOKS-wordpress/README.md [hunk 23]
```diff
@@ -369,10 +599,10 @@ spec:
 Apply via kubectl:
 
 ```console
-kubectl apply -f letsencrypt-issuer.yaml
+kubectl apply -f assets/manifests/letsencrypt-issuer.yaml
 ```
 
-To secure the traffic for WordPress, open the helm file `(value.yml)` created earlier and add the following helm values at the end:
+To secure WordPress traffic, open the helm values file `(values.yaml)` created earlier, and add the following settings at the end:
```
Comment (v-ctiutiu) [inline DOKS-wordpress/README.md:605]: `(values.yaml)` -> (`values.yaml`)

Comment (v-ctiutiu) [inline DOKS-wordpress/README.md:605]: (`values.yaml`)

File: DOKS-wordpress/README.md [hunk 24]
```diff
@@ -266,6 +407,94 @@ NAME                                   DESIRED   CURRENT   READY   AGE
 replicaset.apps/wordpress-6f55c9ffbd   1         1         1       23h
 ```
 
+Verify the PVCs created under the wordpress namespace and associated OpenEBS volume under the openebs namespace:
```
Comment (v-ctiutiu) [inline DOKS-wordpress/README.md:410]: Maybe:

>Verify the PVCs created under the wordpress namespace, and associated OpenEBS volume under the openebs namespace:

Comment (v-ctiutiu) [inline DOKS-wordpress/README.md:410]: >Verify the PVCs created under the wordpress namespace, and associated OpenEBS volume under the openebs namespace:

File: DOKS-wordpress/README.md [hunk 25]
```diff
@@ -266,6 +407,94 @@ NAME                                   DESIRED   CURRENT   READY   AGE
...
+
+The ouput looks similar to:
+
+```text
+NAME                                       CAPACITY   ACCESS MODES   RECLAIM POLICY   STATUS   CLAIM                                                  STORAGECLASS       REASON   AGE
+pvc-b505c0af-e6ab-4623-8ad1-1bad784261d5   5Gi        RWX            Delete           Bound    wordpress/wordpress                                    rwx-storage                 23m
+pvc-d3f0c597-69ba-4710-bd7d-ed29ce41ce04   5Gi        RWO            Delete           Bound    openebs/nfs-pvc-b505c0af-e6ab-4623-8ad1-1bad784261d5   do-block-storage            23m
+```
+
+You can also create additional pod replicas to demonstrate the capabilities of the NFS provisioner by opening the `values.yaml` file and adding the: `replicaCount` line set to the number of desired replicas.
```
Comment (v-ctiutiu) [inline DOKS-wordpress/README.md:438]: Maybe:

>You can also create additional pods to demonstrate the capabilities of the NFS provisioner by opening the `values.yaml` file, and add the `replicaCount` line set to the number of desired replicas.

Comment (v-ctiutiu) [inline DOKS-wordpress/README.md:438]: Maybe:

>You can also create additional pods to demonstrate the capabilities of the NFS provisioner by opening the `values.yaml` file, and add the `replicaCount` line set to the number of desired replicas.

File: DOKS-wordpress/README.md [hunk 26]
```diff
@@ -266,6 +407,94 @@ NAME                                   DESIRED   CURRENT   READY   AGE
...
+
+```yaml
+...
+replicaCount: 3
+...
+```
+
+Apply the changes by running:
```
Comment (v-ctiutiu) [inline DOKS-wordpress/README.md:446]: Maybe (gives more context to the reader):

> Apply changes using the `helm upgrade` command:

Comment (v-ctiutiu) [inline DOKS-wordpress/README.md:446]: Maybe (gives more context to the reader):

> Apply changes using the` helm upgrade` command:

**Review by v-ctiutiu — CHANGES_REQUESTED** [review]

**Review by v-ctiutiu — CHANGES_REQUESTED** [review]

**Review by v-ctiutiu — APPROVED** [review]
Looks good to me.

**Comment by bikram20** [comment]
Nice work, @v-bpastiu!!
